OsmAnd/src/net/osmand/plus/views/mapwidgets/widgets/routeinfo/RouteInfoWidget.java

```diff
@@ -0,0 +1,329 @@
+		inflater.inflate(cachedContentLayoutId, container);
+		collectViews();
+		if (textState != null) {
+			view.setBackgroundResource(textState.widgetBackgroundId);
+			updateNavigationButtonBg();
+		}
+		updateWidgetView();
+
+		buttonTappableArea.setOnClickListener(v -> mapActivity.getMapActions().doRoute());
+
+		view.setOnLongClickListener(v -> {
+			WidgetsContextMenu.showMenu(view, mapActivity, widgetType, customId, null, true, nightMode);
+			return true;
+		});
+	}
+
+	private void collectViews() {
+		buttonTappableArea = view.findViewById(R.id.button_tappable_area);
+		buttonBody = view.findViewById(R.id.button_body);
+		blocksDivider = view.findViewById(R.id.blocks_divider);
+
+		// Initialization of primary block elements
+		tvPrimaryValue1 = view.findViewById(R.id.primary_value_1);
+		tvSecondaryValue1 = view.findViewById(R.id.secondary_value_1);
+		tvTertiaryValue1 = view.findViewById(R.id.tertiary_value_1);
+
+		// Initialization of secondary block elements
+		secondaryBlock = view.findViewById(R.id.secondary_block);
+		tvPrimaryValue2 = view.findViewById(R.id.primary_value_2);
+		tvSecondaryValue2 = view.findViewById(R.id.secondary_value_2);
+		tvTertiaryValue2 = view.findViewById(R.id.tertiary_value_2);
+	}
+
+	@Override
+	public void updateValueAlign(boolean fullRow) {
+	}
+
+	@Override
+	public void updateFullRowState(boolean fullRow) {
+		if (isFullRow != fullRow) {
+			isFullRow = fullRow;
+			recreateView();
+			if (textState != null) {
+				updateColors(textState);
+			}
+		}
+	}
+
+	private void updateNavigationButtonBg() {
+		int color = ColorUtilities.getSecondaryActiveColor(app, nightMode);
+		Drawable normal = UiUtilities.createTintedDrawable(app, R.drawable.rectangle_rounded_small, color);
+
+		int rippleDrawableId = nightMode ? R.drawable.ripple_solid_dark_3dp : R.drawable.ripple_solid_light_3dp;
+		Drawable selected = AppCompatResources.getDrawable(app, rippleDrawableId);
+
+		Drawable drawable = UiUtilities.getLayeredIcon(normal, selected);
+		AndroidUtils.setBackground(buttonBody, drawable);
+	}
+
+	@Override
+	public void updateInfo(@Nullable DrawSettings drawSettings) {
+		updateInfoInternal();
+	}
+
+	private void updateInfoInternal() {
+		calculatedRouteInfo = calculator.calculateRouteInformation();
+		if (cachedContentLayoutId != getContentLayoutId()) {
+			// Recreating the widget is necessary because small widget size uses
+			// different layouts depending on the number of route points.
+			recreateView();
+			return;
+		}
+		boolean shouldHideTopWidgets = mapActivity.getWidgetsVisibilityHelper().shouldHideVerticalWidgets();
+		boolean typeAllowed = widgetType != null && widgetType.isAllowed();
+		boolean visible = typeAllowed && !shouldHideTopWidgets && hasInfoToDisplay();
+		updateVisibility(visible);
+		if (typeAllowed && !shouldHideTopWidgets) {
+			updateNavigationInfo();
+		}
+	}
+
+	private void updateNavigationInfo() {
+		if (Algorithms.isEmpty(calculatedRouteInfo)) {
```

Comment:
We could check if calculatedRouteInfo changes & update UI after it, try to avoid spaming UI updates with the same data